pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,262 @@
+                ids, _, _ = self.grouped.grouper.group_info
+
+                # Drop NA values in grouping
+                mask &= ids != -1
+
+                result_index = self.grouped.grouper.result_index
+                result.index = result_index[ids[mask]]
+
+                if not self.grouped.observed and isinstance(
+                    result_index, CategoricalIndex
+                ):
+                    result = result.reindex(result_index)
+
+                result = self.grouped._reindex_output(result)
+                if self.grouped.sort:
+                    result = result.sort_index()
+
+            return result
+
+    def _handle_int(self, arg):
+        if arg >= 0:
+            return self._ascending_count == arg
+
+        else:
+            return self._descending_count == (-arg - 1)
+
+    def _handle_list(self, args):
+        positive = [arg for arg in args if arg >= 0]
+        negative = [-arg - 1 for arg in args if arg < 0]
+
+        mask = False
+
+        if positive:
+            mask |= np.isin(self._ascending_count, positive)
+
+        if negative:
+            mask |= np.isin(self._descending_count, negative)
+
+        return mask
+
+    def _handle_tuple(self, args):
+        mask = False
+
+        for arg in args:
+            if isinstance(arg, int):
+                mask |= self._handle_int(arg)
+
+            elif isinstance(arg, slice):
+                mask |= self._handle_slice(arg)
+
+            else:
+                raise ValueError(
+                    f"Invalid argument {type(arg)}. Should be int or slice."
+                )
+
+        return mask
+
+    def _handle_slice(self, arg):
+        start = arg.start
+        stop = arg.stop
+        step = arg.step
+
+        if step is not None and step < 0:
+            raise ValueError(f"Invalid step {step}. Must be non-negative")
+
+        mask = True
+
+        if step is None:
+            step = 1
+
+        if start is None:
+            if step > 1:
+                mask &= self._ascending_count % step == 0
+
+        else:
+            if start >= 0:
+                mask &= self._ascending_count >= start
+
+                if step > 1:
+                    mask &= (self._ascending_count - start) % step == 0
+
+            else:
+                mask &= self._descending_count < -start
+
+                offset_array = self._descending_count + start + 1
+                limit_array = (
+                    self._ascending_count + self._descending_count + (start + 1)
+                ) < 0
+                offset_array = np.where(
+                    limit_array, self._ascending_count, offset_array
+                )
+
+                mask &= offset_array % step == 0
+
+        if stop is not None:
+            if stop >= 0:
+                mask &= self._ascending_count < stop
+
+            else:
+                mask &= self._descending_count >= -stop
+
+        return mask
+
+    @property
```

Comment:
rather than do this can you simply compute in `__init__` ?  alt these should be reset during __init__